test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,394 @@
+    CAP_CHANNELS=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d ch_max -v | xargs)
+    CAP_HW=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d dev -v | xargs)
+    PLAY_HW=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:playback" -d dev -v | xargs)
+    CAP_PGA=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:capture" -d pga -v | xargs)
+    PLAY_PGA=$($TPLGREADER "$TPLG" -f "id:$PCM_ID & type:playback" -d pga -v | xargs)
+    export CAP_CHANNELS
+    export CAP_HW
+    export PLAY_HW
+
+    dlogi Test uses "$CAP_CHANNELS" channels
+    dlogi Playback device is "$PLAY_HW"
+    dlogi Playback PGA is "$PLAY_PGA"
+    dlogi Capture device is "$CAP_HW"
+    dlogi Capture PGA is "$CAP_PGA"
+
+    # Find amixer controls for capture, error if more than one PGA
+    numpga=${#CAP_PGA[@]}
+    if [ "$numpga" != 1 ]; then
+	dloge Error: more than one capture PGA found.
+	exit 1
+    fi
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Volume)
+    search="name="
+    CAP_VOLUME=${tmp#*$search}
+    export CAP_VOLUME
+
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Switch)
+    search="name="
+    CAP_SWITCH=${tmp#*$search}
+    export CAP_SWITCH
+
+    dlogi Capture Volume name is "$CAP_VOLUME"
+    dlogi Capture Volume name is "$CAP_SWITCH"
+
+    # Check needed controls
+    amixer cget name="$CAP_VOLUME"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: failed capture volume get command"
+	exit $ret
+    fi
+
+    amixer cget name="$CAP_SWITCH"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: failed capture switch get command"
+	exit $ret
+    fi
+
+    for pga in $PLAY_PGA
+    do
+	tmp=$(amixer controls | grep "$pga" | grep Volume)
+	search="name="
+	play_volume=${tmp#*$search}
+	dlogi Set "$play_volume" to 100%
+	amixer cset name="$play_volume" 100%
+	ret=$?
+	if [ $ret -ne 0 ]; then
+	    dloge "Error: failed play volume set command"
+	    exit $ret
+	fi
+    done
+}
+
+generate_sine () {
+
+    dlogi Creating sine wave file
+    octave --silent --no-gui --eval "check_volume_levels('generate', '$APLAY_WAV')"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: failed sine wave generate."
+	exit $ret
+    fi
+}
+
+measure_levels () {
+
+    dlogi Measuring volume gains
+    octave --silent --no-gui --eval "check_volume_levels('measure', '$ARECORD_WAV1', '$ARECORD_WAV2', '$ARECORD_WAV3')"
+    ret=$?
+    if [ $ret -ne 0 ]; then
+	dloge "Error: Failed one or more tests in volume levels check."
+	dloge "Please inspect files: $ARECORD_WAV1, $ARECORD_WAV1, and $ARECORD_WAV1."
+	exit $ret
+    fi
+}
+
+#
+# Start test
+#
+
+# Preparations
+generate_sine
+do_preparations
+
+# Start sine playback
+aplay_in_background "$APLAY_WAV"
+aplay=$!
+wait
+
+# Do capture tests
+run_test_1
+run_test_2
+run_test_3
+
+# Stop sine playback and cleanup
+wait
+kill $aplay
+
+# Measure
+measure_levels
+ret=$?
+if [ $ret -eq 0 ]; then
+    rm -f "$APLAY_WAV"
+    rm -f "$ARECORD_WAV1"
```

Comment:
You need this for compability with `set -e` (which is missing right now)
```shell
if measure_levels; then
 rm -f "$APLAY_WAV" "$ARECORD_WAV1"
fi
```
Not sure how this is not triggering https://github.com/koalaman/shellcheck/wiki/SC2181 ?!

Reply:
OK, there's a bunch of them.

Reply:
Done.